### ISSUE TITLE
H2 native raster Functions

### ISSUE THREAD

**Issue by nicolas-f** [orbisgis/h2gis#586]
- [x] raster::geometry 
 - Convex hull of the raster (taking account of rotation)
- [x] ST_MetaData( RASTER ) ->ARRAY (Coordinate X, Coordinate Y, Width, Height, ScaleX, ScaleY, SkewX, SkewY, SRID, bands count)
 - Get WKB metadata
- [x] ST_RasterFromImage( BLOB, upperLeftX, upperLeftY, scaleX, scaleY, skewX,
skewY, srid)
 - Convert image stream into Raster using ImageIO (JAI can extend supported formats)
- [ ] ST_ImageFromRaster( RASTER, FormatSuffix )
 - FormatSuffix ex: png tiff
 - ImageIO support also MimeType, but this is less user friendly.

**Comment by ebocher** [comment]
ST_ImageFromRaster must detect the format suffix no ?

**Comment by nicolas-f** [comment]
The format of Raster is WKB Raster. This function must translate into another format, defined by the user.

**Comment by ebocher** [comment]
So this functions is used to convert ?

**Comment by nicolas-f** [comment]
Yes this is the inverse of ST_RasterFromImage. So you should call something like that:
```
SELECT ST_ImageFromRaster(ST_RASTERFROMIMAGE(FILE_READ('/home/user/aerial.png'), 150, 200, 1, 1, 0, 0, 4326), 'bmp') bmpData;
```

**Comment by ebocher** [comment]
Perfect

**Comment by nicolas-f** [comment]
Added ImageIO drivers

**Comment by nicolas-f** [comment]
Added  ST_ImageFromRaster

**Comment by nicolas-f** [comment]
Maybe implement  FILE_WRITE( lob, String ), to write image obtained through ImageFromRaster ?

**Comment by ebocher** [comment]
Yes rigth

**Comment by nicolas-f** [comment]
Add ST_MAKEEMPTYRASTER

**Comment by nicolas-f** [comment]
Raster.toString() is converted into human readable metadata:
```w:10 h:10 bands:3 srid:27572 x:0.00000 y:0.00000 scalex:1.00000 scaley:-1.00000 skewx:0.00000 skewy:0.00000```

**Comment by gpetit** [comment]
@nicolas-f 
Maybe its more clear to write ST_MakeEmptyRaster ?

**Comment by nicolas-f** [comment]
ok..

**Comment by nicolas-f** [comment]
Add ST_BandMetaData

**Comment by nicolas-f** [comment]
Currently WKB Image parsing is slow because there is absolutely no pixel cache. Here a sample of what is done currently:

**I want RGB of pixel 500**
red = Blob.getInputStream().skip(adressOfRedBandPixelZero)
green = stream.skip(adressOfGreenPixelZero)
blue = stream.skip(adressOfBluePixelZero)

**I want RGB of Pixel 501**
The problem here is the stream is in the blue band area. We have to reopen the stream and skip to the red band for pixel 501.

# One solution

The class WKBRasterDataBuffer should cache the bands values for a row. The goal is to reset the minimum of times the stream and reduce the seek amount.

**Comment by nicolas-f** [comment]
Add 
ST_WorldToRasterCoord
 ST_WorldToRasterCoordX
 ST_WorldToRasterCoordY

**Comment by ebocher** [comment]
To be consistent with geometry drivers, ST_RasterFromFile function will be renamed by ST_WorldImageRead (eg SHPRead) and a new function ST_WorldImageWrite will be added.

**Comment by ebocher** [comment]
Note : I don't want to use raster because it's a type (a category of datatype) and not a format like worldfile, geotiff.

**Comment by nicolas-f** [comment]
what about ST_RasterFromWorldFile ?

**Comment by ebocher** [comment]
No because we must also write st_geometryfromfile ;-)

**Comment by ebocher** [comment]
ST_WorldFileImageWrite will have the following signature 

```sql
ST_WorldFileImageWrite(RASTER, '/tmp/toto.png')
```

Usages

```sql
SELECT ST_WorldFileImageWrite(the_raster, '/tmp/toto.png') from myRasterTable limit 1;
```


```sql
SELECT ST_WorldFileImageWrite(the_raster,concat( concat( '/tmp/toto_', pkValue), '.png')) from myRasterTable;
```

**Comment by gpetit** [comment]
@nico any input about this feature ? We are looking forward using raster in H2GIS ;-)

**Comment by ebocher** [comment]
Note : The TIFF format will be integrated in JAVA 9
http://openjdk.java.net/jeps/262

**Comment by nicolas-f** [comment]
Nice.

@gpetit 
 I've merged the raster branch with the last h2 update. I have to write documentation and then create a pull request to h2 database.

**Comment by ebocher** [comment]
+1
We can help you to write the documentation.

**Comment by nicolas-f** [comment]
The branch is here
https://github.com/nicolas-f/h2database/tree/georaster_patch

Off db raster is still not supported yet. But band metadata can already store the path of the external image.

**Comment by ebocher** [comment]
Add a unit test ?
SELECT FILE_WRITE(ST_IMAGEFROMRASTER(data, 'png'), '/tmp/test.png')  FROM test"

**Comment by ebocher** [comment]
Test done

**Comment by ebocher** [comment]
A good day 
https://github.com/h2database/h2database/pull/945
